[2주차] 임규영 숫자 야구 게임 MVC 패턴 적용 제출합니다.

gxuoo opened the pull request:
안녕하세요, 김민석 리뷰어님!

그리디 프론트엔드 2기 임규영입니다. 이번 미션을 통해 처음으로 인사드립니다! 만나서 반갑습니다!! 🙇🏻‍♂️

이번 미션은 1주차 때 만들었던 숫자 야구 게임을 MVC 패턴을 적용시켜 `refactoring` 하는 미션이었습니다. 기존의 `index.js`, `BaseballGame.js` 2가지로 구성했던 게임 로직을 `Model`, `View`, `Controller`로 구분하여 메서드와 클래스, 변수를 MVC 패턴에 알맞게 구분했습니다.

프로젝트 구조를 먼저 설명드린 뒤, `refactoring` 하면서 생긴 고민했던 점, 중점적으로 봐주셨으면 하는 점에 대해서 말씀 드리겠습니다.

## 🗂️ 프로젝트 구조
### index.html
- `index.js`을 실행합니다.

### index.js
- `Controller`인스턴스를 생성하여 게임을 실행합니다.

### NumberBaseballGameController.js
- `Model`, `View`인스턴스를 생성하여 상호 간의 데이터를 주고 받아 필요한 메서드를 호출하는 역할을 수행합니다.
- `View`에서 `HTML Element` 일부를 받아와 `eventListener`를 추가하여 사용자 입력에 반응합니다.
- 게임 시작 및 재시작을 관리합니다.
    - `Model`에게 사용자 입력 값을 전달합니다.
    - `View`에게 `Model`에게서 받은 결과를 전달합니다.

### NumberBaseballGameModel.js
- `Controller`에게 받은 사용자 입력에 대한 유효성을 검증한 뒤에 게임 진행(`Strike`, `Ball`판정)을 수행하고 이에 대한 결과 문자열을 `Controller`에게 전달합니다.
- 정답 숫자 문자열을 생성 및 관리합니다.


### NumberBaseballGameView.js
- `HTML Elements`를 `private`변수에 받아와 관리합니다.
- `Controller`에게 받은 `Model`에서 처리한 결과 값을 화면에 출력합니다.
    - 게임 결과 문자열, 오류 메시지
- `<button>`, `<Input>`태그의 활성 및 비활성화를 관리합니다.
- `<Input>`에 `focus`를 주어 불필요한 마우스 클릭을 하지 않도록 구현했습니다.

### constants.js
- 반복되는 숫자, 문자열에 대해 의미있는 이름을 가진 `상수`로 선언하여 모아 놓은 파일입니다.


## 🤔 고민했던 점
### 1. `View`의 데이터 조작
- 위 캡쳐본은 이번 미션에서 `View`의 기능 요구 사항입니다. 여기서 “`View`에서 데이터의 조작이 일어나선 안된다.” 라는 문구 중, `View`에서 데이터의 조작은 `Model`이나 `Controller`의 데이터 조작을 해서는 안된다는 말인지 궁금합니다.
    - 제가 판단하기로는 `View`에서도 화면에 출력을 위한 `View`가 가지고 있는 데이터의 조작은 필요하다고 생각하여 해당 문구는 `Model` 혹은 `Controller`의 데이터 조작이 일어나서는 안 된다고 이해했습니다.


## ❗️중점적으로 봐주셨으면 하는 점
1. MVC 패턴을 잘 적용시켰는지
2. 변수, 메서드의 네이밍이 말하고자 하는 내용을 적절하게 나타내는지
3. 불필요한 메서드의 호출하는 구조가 존재하는지
4. 커밋을 단위가 적절한지

## 💻 npm run test 결과
![image](https://github.com/user-attachments/assets/1f23bf5a-64fe-4d9a-9fa1-fb5004e4ca62)


-----
이상으로 이번 2주차 미션 PR를 제출합니다. PR에 대해 보완할 부분이나 불필요한 내용이 있다면 수정하겠습니다. 잘 부탁드립니다!!

gxuoo committed:
docs: README.md에 MVC 추가

1주차에 구현했던 기능들을 MVC 패턴에 알맞게 분류한 내용을 README.md에 추가했습니다.

gxuoo committed:
fix: HTML 형식 수정

HTML 일부 태그를 게임 결과 출력 형식을 맞추기 위해 수정했습니다.

gxuoo committed:
refactor: MVC 패턴 적용을 위한 파일 생성

MVC 패턴 적용을 위한 Model, View, Controller, index.js 총 4가지 파일을 생성 및 수정했습니다.

gxuoo committed:
reafactor: Controller Refactoring

Controller에 해당되는 메서드를 모아 하나의 Claas로 refactoring 했습니다.

기존에 eventListener를 add하는 방식이 리팩토링 과정에서 this 함수가 클래스 인스턴스를 가르키지 않고 HTML Element를 가리키는 문제가 있어, 화살표 함수로 변경했습니다.

gxuoo committed:
feat: 상수 파일 추가

자주 사용되는 숫자, 문자열을 모아놓은 상수 파일입니다.

gxuoo committed:
refactor: Model Refactoring

Model에 해당되는 메서드와 변수를 하나의 클래스로 만들었고, 기존의 BaseballGame 클래스에서 사용자 입력에 대한 검증 메서드를 추가했습니다.

gxuoo committed:
refactor: View Refactoring

View에 해당되는 메서드와 변수를 하나의 클래스로 모았습니다.

HTML Elemnet를 private 변수로 관리하며, 오류 메시지와 결과 메시지 화면에 보여지는 <button>, <input> 태그의 활성 및 비활성화를 담당합니다.

gxuoo committed:
fix: 오타 수정

메서드 네이밍, 클래스 명, 파일 명에 오타를 수정했습니다.

gxuoo committed:
fix: 불필요한 메서드 제거 및 오타 수정

사용하지 않는 메서드를 제거했고, 파일명 오타를 수정했습니다.

gxuoo committed:
fix: 메서드 호출 위치 변경

Controller에서 호출하지 않고, View 생성자에서 호출해도 된다고 판단하여, 메서드 호출 위치를 변경했습니다.

gxuoo committed:
fix: 메서드 네이밍 수정

각 메서드의 역할을 나타낼 수 있는 네이밍으로 수정했습니다.

gxuoo committed:
fix: 메서드 호출 위치 변경, 오타 수정

이전 커밋에서 메서드 호출 위치가 변경되어 이를 수정했고, 오타가 있어 수정했습니다.

shackstack reviewed (changes requested):
규영님 반갑습니다~
연차쓰고 해외여행을 좀 다녀오느라 리뷰가 조금 늦었네요.

피드백을 드리자면.. view는 화면을 담당, model은 비즈니스 로직을 담당, controller는 중개역할 이렇게 역할을 잘 분리하신 것 같아요. 네이밍도 괜찮습니다. 불필요한 메서드를 호출하는 부분은 없어보이나, 외부에서 사용하지 않는 메서드 관리를 조금더 하시면 좋을 것 같아요. 커밋의 경우엔 단위도 적절하고, 커밋에 디스크립션까지 달아주는 부분 좋았습니다만 디스크립션은 `File Changed` 에도 달 수 있으니, 어떤방법이 더 좋을지 고민해보셔도 좋을 것 같아요. 그리고 굳이 설명을 달지 않아도, 코드로 설명된다면 덜어내셔도 괜찮을 것 같습니다. 그리고 프로젝트 구조 문서화 깔끔하게 하셔서 보기 너무 편했어요.

### 질문에 대한 답변
> 1. `View`의 데이터 조작
위 캡쳐본은 이번 미션에서 `View`의 기능 요구 사항입니다. 여기서 “`View`에서 데이터의 조작이 일어나선 안된다.” 라는 문구 중, `View`에서 데이터의 조작은 `Model`이나 `Controller`의 데이터 조작을 해서는 안된다는 말인지 궁금합니다. 제가 판단하기로는 `View`에서도 화면에 출력을 위한 `View`가 가지고 있는 데이터의 조작은 필요하다고 생각하여 해당 문구는 `Model` 혹은 `Controller`의 데이터 조작이 일어나서는 안 된다고 이해했습니다.

우선 캡쳐본이 보이지 않습니다... 대충 View가 데이터를 조작해도 되는지에 대한 질문인 것 같은데, 안될건 없습니다. 불가피하다면요. 다만 거기에 따라 발생할 수 있는 유지보수성에 대한 어려움은 감당하셔야겠죠. MVC가 생겨나게된 배경에 대해서 조금더 고민해보시면 답을 찾으실 수 있으실 거라 봐요.

코멘트 확인하시고 다시 PR 요청주세요~ 화이팅입니다.

gxuoo committed:
fix: 변수 및 함수 네이밍 수정, 위치 수정

camelCase 통일, 함수 네이밍 시작을 동사로 통일, 무엇을 나타내는지 명확하게 표시, 함수 위치 수정

gxuoo committed:
fix: HTML 태그 출력 형식 변경

기존의 방식 대신 DocumentFragment 객체와 Range 객체를 사용하여 HTML 태그를 동적으로 생성했습니다.

gxuoo committed:
fix: 사용자 입력 유효성 검증 함수의 기능 분리

사용자 검증 유효성 검증 함수의 역할을 분리, 오류 메시지를 생성하는 함수를 추가했고 외부에서 사용되지 않는 함수를 private 선언했습니다.

gxuoo reviewed (commented):
리뷰어님의 리뷰를 반영해서 코드를 수정해봤습니다. 민석님의 귀한 시간을 제 코드 리뷰에 할애해주셔서 감사드립니다! 

다만, 이번 커밋을 진행하면서 커밋 단위를 잘못 올린 것 같아 이전 커밋으로 `--hard` 옵션으로 `reset` 진행하여 새로 커밋을 올렸는데 해당 방식이 잘못 됐다면 제가 git에 대해서 조금 더 공부해보고 커밋과 푸쉬에 대한 실수가 없도록 노력해보겠습니다.

gxuoo committed:
Merge branch 'gxuoo' into gxuoo

gxuoo committed:
refactor: 리팩토링 과정 중 사용하지 않은 파일 삭제

shackstack commented on src/numberBaseballGameController.js:
```diff
@@ -0,0 +1,43 @@
+        const user = userInput.value.trim();
+        const messages = [];
+
+        if (!this.#numberBaseballGameModel.checkUserInputValidation(user, messages)) {
```
음... pushErrorMessageToArray를 분리한다고 checkUserInputValidation의 역할이 줄어든다고 볼 수 있을까요?
여전히 messages를 업데이트하고, 상태를 return 하고 있습니다.

`checkUserInputValidation`가 
```
{ isValid: boolean. messages: string[]} 
```
를 리턴하게되면, `validation 결과를 내뱉는다` 라는 측면에서 한가지 역할만 한다고 볼 수 있지 않을까요?

shackstack commented on src/numberBaseballGameView.js:
```diff
@@ -0,0 +1,77 @@
+        this.#userInput.focus();
+    }
+
+    alertErrorMessage(messages) {
```
alertErrorMessage라면 string형태의 message를 입력받아야할 것 같아요.
인자와 함수네이밍을 맞추면 좋을 것 같습니다.

shackstack commented on src/numberBaseballGameView.js:
```diff
@@ -0,0 +1,77 @@
+        const template = `
+        <strong>🎉정답을 맞추셨습니다🎉</strong><br><br>
+        게임을 다시 시작하시겠습니까?<br><br>
+        `;
```
👍

shackstack commented on src/numberBaseballGameView.js:
```diff
@@ -0,0 +1,77 @@
+        게임을 다시 시작하시겠습니까?<br><br>
+        `;
+
+        const fragment = document.createRange().createContextualFragment(`${template}`);
```
```suggestion
        const fragment = document.createRange().createContextualFragment(template);
```

shackstack commented on src/numberBaseballGameModel.js:
```diff
@@ -0,0 +1,66 @@
+        return /^[1-9]{3}$/.test(value);
+    }
+
+    checkDuplicatesInString(value) {
```
<img width="705" alt="스크린샷 2025-03-24 오후 1 48 47" src="https://github.com/user-attachments/assets/0264b990-36a5-4bf9-ba68-4b69de2e7eb6" />

문법적으로 문제가 있어보이지 않나요?

shackstack commented on src/numberBaseballGameView.js:
```diff
@@ -0,0 +1,73 @@
+    showAnswerMessage() {
+        const strongTag = document.createElement('strong');
+        strongTag.textContent = '정답을 맞추셨습니다';
+        this.#resultMessage.append('🎉', strongTag, '🎉', document.createElement('br'), document.createElement('br'), '게임을 다시 시작하시겠습니까?', document.createElement('br'), document.createElement('br'));
```
Javascript 템플릿 리터럴에 HTML 코드를 작성하여 append 해보시라는 의도였습니다.
```
  this.#resultMessage.append(`<div>ABCD</div>`)
```
요런식으로요.

shackstack reviewed (approved):
규영님~ 빠른시간내에 수정요청사항들 반영해주셨군요. 
열정이 대단해보이십니다 👍
다음 미션을 위해 미리 승인해두겠습니다. 
코멘트 확인해주시고 DM이나 코멘트에 답변해주시면 될 것 같아요. 
고생하셨습니다.

gxuoo committed:
fix: 변수, 함수 네이밍 수정, 토글 형태 함수 삭제

boolean 값을 반환하는 함수에 대한 네이밍을 수정, 변수의 뜻을 명확하게 나타내기 위한 네이밍 수정, 불필요한 토글 함수를 삭제했습니다.
